_episodes/32.md

```diff
@@ -0,0 +1,389 @@
+Just as storage was built with fault-tolerance, compute systems like MapReduce
+that processed and transformed data was also [built with fault tolerance in mind](https://static.googleusercontent.com/media/research.google.com/en//archive/mapreduce-osdi04.pdf).
+Apache Hive was syntax and metadata layer that enabled generating MapReduce jobs
+without having to write code. Apache Spark came on the analytics scene
+by [introducing lineage](https://www.usenix.org/system/files/conference/nsdi12/nsdi12-final138.pdf) 
+as a way for engineers to have more control over how and when their datasets
+would be flushed to disk. This technique, while novel, still took a very
+pessimistic view that allowing faults was the worse case scenario to avoid.
+
+When Trino was created, it was designed with speed in mind. Trino creators 
+Martin, Dain, and David chose not to add fault-tolerance to Trino as they
+recognized the tradeoff of fast analytics. Due to the nature of the streaming 
+exchange in Trino all tasks are interconnected. A failure of any task results in
+a query failure. To support long running queries Trino has to be able to 
+tolerate task failures.
+
+Having an all-or-nothing architecture makes it significantly more difficult to 
+tolerate faults, regardless of how rare they are. The likelihood of a failure 
+grows with the time it takes to complete a query. This risk also increases as 
+the resource demands such as memory requirements of a query grow. It’s 
+impossible to know upfront the exact memory requirements that a query will need.
+In addition to increased likelihood of a failure the impact of failing a long 
+running query is much higher as it often results in a significant waste of time
+and resources. 
+
+You may think this is a model that is bound to fail, especially when scaling to
+petabytes of data. On the contrary, Trino's predecessor Presto was commonly used
+to execute batch workloads at this scale at Facebook. Even today, companies like
+[Salesforce](https://medium.com/salesforce-engineering/how-to-etl-at-petabyte-scale-with-trino-5fe8ac134e36), 
+[Doordash](https://www.starburst.io/resources/trino-summit/?wchannelid=2ug6mgs5ao&wmediaid=j1eq196a4y), 
+and many others, use Trino at Petabyte scale to handle ETL workloads. While it 
+is possible, scaling Trino to run petabyte scale ETL pipelines, you really have
+to know what you’re doing.
+
+Resource management is another challenge. Users how to know exactly what 
+resource utilization to expect from a query they submit to properly size the 
+cluster and to avoid resource related failures they have to know how many 
+queries to submit.
+
+In essence, most people avoid using Trino because they lack the understanding of
+how to correctly configure Trino based on the infrastructure they are deploying
+on. 
+
+### What are the limitations of the current architecture?
+
+With current architecture all tasks which compose the query execution are 
```

Comment:
In the current architecture all tasks for processing a specific query are planned upfront.